### PR TITLE
fix(button): fix button layout for icons and hidden text

### PR DESCRIPTION
**Pull request by pvinis**
The type of this PR is: **Bugfix**

This PR resolves [CX-2007]

### Description

With the opportunity to fix this icon layout bug, I made the buttons a bit easier to work with, and added a couple of utils in the process.

- `iconPosition` can be `left-start` now, and that will put the icon all the way at the left side when the button is `block`.
- `MeasuredView` can be used to measure some view's width and height, and that can be used to calculate the longest text for buttons.
- `useOffscreenStyle` (which was kinda invented for the instagram share) is now a hook. It returns a style object that will position the view it is given to, outside the visible area of the screen.
- Some changes in the button stories to make testing easier.

You can see below the way that icons are laid out, and also how the button takes the same width for short and long text.

Screenshot 2021-10-04 at 18 05 11

Also, the onboarding screens are as they are supposed to be.

Screenshot 2021-10-04 at 18 07 33

Also also, the problematic button is fixed.

Screenshot 2021-10-04 at 18 18 55

### PR Checklist (tick all before merging)

- [x] I have included screenshots or videos to illustrate my changes, or I have not changed anything that impacts the UI.
- [x] I have tested my changes on **iOS** and **Android**.
- [x] I have added tests for my changes, or my changes don't require testing, or I have included a link to a separate Jira ticket covering the tests.
- [x] I have added a feature flag, or my changes don't require a feature flag. ([How do I add one?](https://github.com/artsy/eigen/blob/master/docs/developing_a_feature.md))
- [x] I have documented any follow-up work that this PR will require, or it does not require any.
- [x] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/master/docs/adding_state_migrations.md))
- [x] I have added a changelog entry below or my changes do not require one.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

- fix button inner layout - pavlos

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

</details>

[CX-2007]: https://artsyproduct.atlassian.net/browse/CX-2007

### CODE REVIEW

File: src/storybook/decorators.tsx [hunk 1]
```diff
@@ -12,7 +12,9 @@ export const withSafeArea: DecoratorFunction<React.ReactNode> = (story) => (
 )
 
 export const withScreenDimensions: DecoratorFunction<React.ReactNode> = (story) => (
-  <ProvideScreenDimensions>{story()}</ProvideScreenDimensions>
+  <SafeAreaProvider>
```
Comment (pvinis) [inline src/storybook/decorators.tsx:15]: this is needed by the dimensions provider, so lets have it here.

File: src/palette/svgs/LinkIcon.tsx [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
 import { Icon, IconProps, Path } from "./Icon"
 
 export const LinkIcon: React.FC<IconProps> = (props) => {
+  const color = useColor()
   return (
     <Icon {...props} viewBox="0 0 23 23">
       <Path
         d="M18.6761 4.3241C20.3689 6.01656 20.3689 8.77157 18.6761 10.464L16.6993 12.4411C16.7512 11.5994 16.6211 10.7485 16.3081 9.95285L17.2364 9.02434C17.5747 8.68606 17.7857 8.26805 17.8694 7.83031C18.0082 7.10458 17.7972 6.32461 17.2364 5.7638C16.3375 4.8647 14.8749 4.86473 13.9759 5.7638L10.4506 9.28901C9.55152 10.1879 9.5516 11.6506 10.4507 12.5496C10.6636 12.7624 10.9084 12.9239 11.1695 13.0352L9.67317 14.5318C9.43902 14.3759 9.21675 14.1955 9.0107 13.9893C7.31787 12.2968 7.31789 9.54178 9.01072 7.84932L12.536 4.32407C14.2286 2.63105 16.9835 2.63108 18.6761 4.3241Z"
-        fill="#666666"
+        fill={color(props.fill) ?? "#666666"}
```
Comment (pvinis) [inline src/palette/svgs/LinkIcon.tsx:11]: allow for a white link

**Comment by ArtsyOpenSource** [comment]
### This PR contains the following changes:

- Cross-platform user-facing changes (fix button inner layout - pavlos)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4350f3b000771b25495295c13704ce97fbf3a34b
</p>

File: src/palette/elements/Button/Button.tsx [hunk 3]
```diff
@@ -20,7 +21,7 @@ export interface ButtonProps extends BoxProps {
   onPress?: PressableProps["onPress"]
 
   icon?: ReactNode
-  iconPosition?: "left" | "right"
+  iconPosition?: "left" | "left-start" | "right"
```
Comment (pvinis) [inline src/palette/elements/Button/Button.tsx:24]: new option!

File: src/palette/elements/Button/Button.tsx [hunk 4]
```diff
@@ -72,6 +73,8 @@ export const Button: React.FC<ButtonProps> = ({
 }) => {
   const [innerDisplayState, setInnerDisplayState] = useState(DisplayState.Enabled)
 
+  const [longestTextMeasurements, setLongestTextMeasurements] = useState<ViewMeasurements>({ width: 0, height: 0 })
```
Comment (pvinis) [inline src/palette/elements/Button/Button.tsx:76]: here's where we store the measurements of the longest text when it's rendered

File: src/palette/elements/Button/Button.tsx [hunk 5]
```diff
@@ -356,41 +361,6 @@ const useStyleForVariantAndState = (
...
-    opacity={displayState === DisplayState.Loading ? 0 : 1}
-    position="absolute"
-    left={position === "left" ? 0 : undefined}
-    right={position === "right" ? 0 : undefined}
-  >
-    {icon}
-  </Box>
-)
-
-const VisibleTextContainer = styled(Box)`
```
Comment (pvinis) [inline src/palette/elements/Button/Button.tsx:378]: BYE 👋

File: src/palette/elements/Button/Button.tsx [hunk 6]
```diff
@@ -151,37 +154,39 @@ export const Button: React.FC<ButtonProps> = ({
...
-                      position: "absolute",
-                      width: "100%",
                     }}
                     textAlign="center"
                   >
                     {children}
                   </AnimatedText>
-                  {iconPosition === "right" && <IconBox position="right" icon={icon} displayState={displayState} />}
-                </VisibleTextContainer>
+                  {iconPosition === "right" ? icon : null}
```
Comment (pvinis) [inline src/palette/elements/Button/Button.tsx:181]: simpler icon rendering

File: src/palette/elements/Button/Button.tsx [hunk 7]
```diff
@@ -151,37 +154,39 @@ export const Button: React.FC<ButtonProps> = ({
...
+                  {iconPosition === "left" ? icon : null}
+                  <MeasuredView setMeasuredState={setLongestTextMeasurements}>
+                    <Text variant={size === "small" ? "xs" : "sm"} color="red">
+                      {longestText ? longestText : children}
+                    </Text>
+                  </MeasuredView>
                   <AnimatedText
                     variant={size === "small" ? "xs" : "sm"}
                     style={{
+                      width: longestTextMeasurements.width,
```
Comment (pvinis) [inline src/palette/elements/Button/Button.tsx:173]: here's the magic! we give the regular text a width of the longest text. boom!

File: src/palette/elements/Button/Button.tsx [hunk 8]
```diff
@@ -151,37 +154,39 @@ export const Button: React.FC<ButtonProps> = ({
...
-                  {iconPosition === "left" && <IconBox position="left" icon={icon} displayState={displayState} />}
+              <Flex mx={containerSize.mx}>
+                <Flex height="100%" flexDirection="row" alignItems="center" justifyContent="center">
+                  {iconPosition === "left-start" ? (
+                    <Box position="absolute" left={0}>
+                      {icon}
+                    </Box>
+                  ) : null}
+                  {iconPosition === "left" ? icon : null}
+                  <MeasuredView setMeasuredState={setLongestTextMeasurements}>
```
Comment (pvinis) [inline src/palette/elements/Button/Button.tsx:165]: we render an offscreen view that contains the longest text in the format we want, and we measure it.

**Review by brainbicycle — APPROVED** [review]
Nice!! 🧹

**Pull request merged**